File: windows/tests/Elevate.Core.Tests/LapsedAssignmentTests.cs
using Elevate.Core.Models;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class LapsedAssignmentTests
{
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_000_000);

    private static ActiveAssignment Assignment(AssignmentStatus status, double? endsInSeconds) =>
        new(new RoleKey("i", "t", new EntraDirectoryScope("r", "/")), "a", Now.AddHours(-2),
            endsInSeconds is { } s ? Now.AddSeconds(s) : null, status);

    [Fact]
    public void ActiveLapsesOnceItsEndHasPassed()
    {
        Assignment(AssignmentStatus.Active, -1).HasLapsed(Now).Should().BeTrue();
        Assignment(AssignmentStatus.Active, 0).HasLapsed(Now).Should().BeTrue();
        Assignment(AssignmentStatus.Active, 1).HasLapsed(Now).Should().BeFalse();
    }

    [Fact]
    public void ScheduledLapsesOnceItsEndHasPassed()
    {
        Assignment(AssignmentStatus.Scheduled, -60).HasLapsed(Now).Should().BeTrue();
        Assignment(AssignmentStatus.Scheduled, 60).HasLapsed(Now).Should().BeFalse();
    }

    [Fact]
    public void WithoutAnEndNothingLapses()
    {
        Assignment(AssignmentStatus.Active, null).HasLapsed(Now).Should().BeFalse();
    }

    /// <summary>Requests and failures carry no activation window of their own; the service settles them.</summary>
    [Fact]
    public void RequestsAndFailuresNeverLapse()
    {
        Assignment(AssignmentStatus.PendingApproval, -60).HasLapsed(Now).Should().BeFalse();
        Assignment(AssignmentStatus.PendingProvisioning, -60).HasLapsed(Now).Should().BeFalse();
        Assignment(AssignmentStatus.Failed("x"), -60).HasLapsed(Now).Should().BeFalse();
    }
}

File: macos/Tests/ElevateCoreTests/LapsedAssignmentTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct LapsedAssignmentTests {
    let now = Date(timeIntervalSince1970: 1_000_000)
    func assignment(_ status: ActiveAssignment.Status, endsIn: TimeInterval?) -> ActiveAssignment {
        ActiveAssignment(roleKey: RoleKey(identityId: "i", tenantId: "t", scope: .entraDirectory(roleDefinitionId: "r", directoryScopeId: "/")),
                         assignmentId: "a", startDateTime: now.addingTimeInterval(-7200),
                         endDateTime: endsIn.map { now.addingTimeInterval($0) }, status: status)
    }

    @Test func activeLapsesOnceItsEndHasPassed() {
        #expect(assignment(.active, endsIn: -1).hasLapsed(at: now))
        #expect(assignment(.active, endsIn: 0).hasLapsed(at: now))
        #expect(!assignment(.active, endsIn: 1).hasLapsed(at: now))
    }

    @Test func scheduledLapsesOnceItsEndHasPassed() {
        #expect(assignment(.scheduled, endsIn: -60).hasLapsed(at: now))
        #expect(!assignment(.scheduled, endsIn: 60).hasLapsed(at: now))
    }

    @Test func withoutAnEndNothingLapses() {
        #expect(!assignment(.active, endsIn: nil).hasLapsed(at: now))
    }

    /// Requests and failures carry no activation window of their own; the service settles them.
    @Test func requestsAndFailuresNeverLapse() {
        #expect(!assignment(.pendingApproval, endsIn: -60).hasLapsed(at: now))
        #expect(!assignment(.pendingProvisioning, endsIn: -60).hasLapsed(at: now))
        #expect(!assignment(.failed("x"), endsIn: -60).hasLapsed(at: now))
    }
}
